file-uploader/file/file_processor.go

```diff
@@ -0,0 +1,58 @@
+package file
+
+import (
+	"bufio"
+	"cloud.google.com/go/pubsub"
+	"context"
+	log "github.com/sirupsen/logrus"
+	"mime/multipart"
+	"github.com/ONSdigital/ras-rm-sample/file-uploader/config"
+)
+
+type FileProcessor struct {
+	Config config.Config
+	Client *pubsub.Client
+	Ctx context.Context
+}
+
+func (f *FileProcessor) ChunkCsv(file multipart.File, handler *multipart.FileHeader) {
+	log.WithField("filename", handler.Filename).
+		WithField("filesize", handler.Size).
+		WithField("MIMEHeader", handler.Header).
+		Info("File uploaded")
+	f.Publish(bufio.NewScanner(file))
+}
+
+func (f *FileProcessor) Publish(scanner *bufio.Scanner) int {
+	topic := f.Client.Topic(f.Config.Pubsub.TopicId)
+	var errorCount = 0
+	for scanner.Scan() {
+		line := scanner.Text()
+		log.WithField("line", line).
+			Debug("Publishing csv line")
+
+		errorChannel := make(chan error, 0)
+
+		go func(line string, topic *pubsub.Topic) {
+			id, err := topic.Publish(f.Ctx, &pubsub.Message{
+				Data: []byte(line),
+			}).Get(f.Ctx)
+			log.WithField("line", line).
+				WithField("messageId", id).
+				Debug("csv line acknowledged")
+			errorChannel <- err
+		}(line, topic)
+
+		err := <- errorChannel
```

Comment:
I may be wrong, but won't this loop launch a go rountine to process a line in the CSV and then wait for a the channel to receive an error (or nil). Is that correct?

Reply:
This is where I show my lack of understanding with go. I wasnt aware that that the channel would block. I only put the go routine in because Publish.Get is a blocking action but if the error channel also blocks then theres no point in having the go routine there.
What would you recommend? I only wanted to make a running count of errors.

Reply:
I'd be tempted to use a wait group, spin up all the goroutines, exit the loop and wait. Then read from the channel at the once they're all done to collect the errors

There's a nice simple example here
https://gobyexample.com/waitgroups

Reply:
and you can do a range over the channel like this:

https://gobyexample.com/range-over-channels

Reply:
actually you could skip the channel and use a mutex lock to append to a list of errors directly in the go rountine itself